spec/lib/rmt/cli/products_spec.rb

```diff
@@ -203,38 +286,38 @@
         end
         let(:all_products) { [product] + extensions + non_free_extensions }
         let(:products_to_enable) { all_products - non_free_extensions }
-        let(:repo_count) { products_to_enable.inject(0) { |sum, product| sum + product.repositories.where(enabled: true).count } }
-        let(:expected_output) do
-          "The following required extensions for #{product.product_string} have been enabled: #{extensions.pluck(:name).join(', ')}.\n" \
-          "#{repo_count} repo(s) successfully enabled.\n"
-        end
+        let(:repos) { products_to_enable.flat_map { |product| product.repositories.where(enabled: true) } }
 
         it 'enables product and recommended products repositories' do
+          expect { described_class.start(argv) }.to output(expected_output).to_stdout.and output('').to_stderr
           products_to_enable.flat_map(&:repositories).each do |repository|
             expect(repository.mirroring_enabled).to eq(repository.enabled)
           end
         end
 
         it 'does not enable non-free extensions' do
+          expect { described_class.start(argv) }.to output(expected_output).to_stdout.and output('').to_stderr
           non_free_extensions.flat_map(&:repositories).each do |repository|
             expect(repository.mirroring_enabled).to be_falsey
           end
         end
 
         it 'has more repositories than the base product' do
-          expect(repo_count).to be > product.repositories.where(enabled: true).count
+          expect { described_class.start(argv) }.to output(expected_output).to_stdout.and output('').to_stderr
+          expect(repos.count).to be > product.repositories.where(enabled: true).count
         end
       end
     end
 
-    context 'by wrong product ID' do
-      let(:false_id) { (product.id + 1).to_s }
-      let(:argv) { ['enable', false_id] }
-      let(:expected_output) { "Product by id \"#{false_id}\" not found.\n" }
+    context 'by wrong product string' do
+      let(:target) { 'badproductstring' }
+      let(:argv) { ['enable', target] }
+      let(:expected_stderr) { "Product(s) #{target} could not be found and were not enabled.\n" }
+      let(:expected_output) { "No product found for target '#{target}'.\n" }
 
       before do
         expect(described_class).to receive(:exit)
```

Comment:
One `before` block missing that needs some love :)

I understand the reasoning though that the `expect(...)...` was put into a before block, as it adds duplication in some points. I still wonder if and how we could clean this up. I wonder whether we can explicitly define a subject with that once in the beginning. If it's too much hassle to refactor in this context it's fine, but it is bugging me.